### PR TITLE
Add clippy into the known `cfg` list

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
In clippy, we are removing the `feature = "cargo-clippy"` cfg to replace it with `clippy` in https://github.com/rust-lang/rust-clippy/pull/12292. But for it to work, we need to declare `clippy` as cfg. It makes it more coherent with other existing tools like rustdoc.

cc @flip1995

### CODE REVIEW

File: compiler/rustc_session/src/config.rs [hunk 1]
```diff
@@ -1441,6 +1441,7 @@ impl CheckCfg {
         // These three are never set by rustc, but we set them anyway: they
         // should not trigger a lint because `cargo doc`, `cargo test`, and
         // `cargo miri run` (respectively) can set them.
```
Comment (flip1995) [inline compiler/rustc_session/src/config.rs:1443]: This comment also needs to be updated.

File: compiler/rustc_session/src/config.rs [hunk 2]
```diff
@@ -1441,6 +1441,7 @@ impl CheckCfg {
...
+        ins!(sym::clippy, no_values);
```
Comment (Urgau) [inline compiler/rustc_session/src/config.rs:1444]: It needs to be added in [tests/ui/check-cfg/well-known-values.rs](https://github.com/rust-lang/rust/blob/4ae1e79876fcee94a56c1edf7c029a677084d7fc/tests/ui/check-cfg/well-known-values.rs).

*You will also probably need to bless the directory*

Comment (Urgau) [inline compiler/rustc_session/src/config.rs:1444]: The check-cfg section of the unstable book also needs updating: https://github.com/rust-lang/rust/blob/4ae1e79876fcee94a56c1edf7c029a677084d7fc/src/doc/unstable-book/src/compiler-flags/check-cfg.md?plain=1#L80-L87

Comment (GuillaumeGomez) [reply compiler/rustc_session/src/config.rs:1444]: Good catch!

File: src/doc/unstable-book/src/compiler-flags/check-cfg.md [hunk 3]
```diff
@@ -81,6 +81,7 @@ As of `2024-01-09T`, the list of known names is as follows:
 
 <!--- See CheckCfg::fill_well_known in compiler/rustc_session/src/config.rs -->
 
+ - `clippy`
```
Comment (Urgau) [inline src/doc/unstable-book/src/compiler-flags/check-cfg.md:84]: Would be great if you could also update the date above to `2024-02-15T`

File: tests/ui/check-cfg/well-known-values.rs [hunk 4]
```diff
@@ -106,4 +106,7 @@ fn unix() {}
 #[cfg(doc)]
 fn doc() {}
 
+#[cfg(clippy)]
+fn clippy() {}
```
Comment (Urgau) [inline tests/ui/check-cfg/well-known-values.rs:110]: Could you also add the cfg above as to check that we only expect `clippy` and nothing else.

```diff
 #[cfg(any(
     // tidy-alphabetical-start
+    clippy = "_UNEXPECTED_VALUE",
+    //~^ WARN unexpected `cfg` condition value
     debug_assertions = "_UNEXPECTED_VALUE",
     //~^ WARN unexpected `cfg` condition value
```

Comment (GuillaumeGomez) [reply tests/ui/check-cfg/well-known-values.rs:110]: Ah sure.

File: src/doc/unstable-book/src/compiler-flags/check-cfg.md (hunk 3 above)
Comment (GuillaumeGomez) [reply src/doc/unstable-book/src/compiler-flags/check-cfg.md:84]: :+1:

**Comment by GuillaumeGomez** [comment]
Updated! Thanks a lot for the reviews!

**Review by Urgau — APPROVED** [review]
Looks good, thanks.

File: compiler/rustc_session/src/config.rs [hunk 5]
```diff
@@ -1438,9 +1438,10 @@ impl CheckCfg {
 
         ins!(sym::debug_assertions, no_values);
 
-        // These three are never set by rustc, but we set them anyway: they
-        // should not trigger a lint because `cargo doc`, `cargo test`, and
-        // `cargo miri run` (respectively) can set them.
+        // These four are never set by rustc, but we set them anyway: they
+        // should not trigger a lint because `cargo doc`, `cargo test`,
+        // `cargo clippy` and `cargo miri run` (respectively) can set them.
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply compiler/rustc_session/src/config.rs:1443]: Ok, let's alpha sort them. :laughing:

File: compiler/rustc_session/src/config.rs (hunk 2 above)
Comment (Nilstrieb) [reply compiler/rustc_session/src/config.rs:1444]: The code here should have a comment mentioning the docs :)

Comment (GuillaumeGomez) [reply compiler/rustc_session/src/config.rs:1444]: Adding it.

**Comment by Nilstrieb** [comment]
Guillaume asked me to review the PR since I commented on it, and it looks good to me.
@bors r=Urgau,Nilstrieb

**Pull request merged**